baseapp/baseapp.go

```diff
@@ -856,7 +864,61 @@ func createEvents(events sdk.Events, msg sdk.Msg) sdk.Events {
 	return sdk.Events{msgEvent}.AppendEvents(events)
 }
 
-// DefaultPrepareProposal returns the default implementation for processing an
+// PrepareProposalVerifyTx performs transaction verification when a proposer is
+// creating a block proposal during PrepareProposal. Any state committed to the
+// PrepareProposal state internally will be discarded. <nil, err> will be
+// returned if the transaction cannot be encoded. <bz, nil> will be returned if
+// the transaction is valid, otherwise <bz, err> will be returned.
+func (app *BaseApp) PrepareProposalVerifyTx(tx sdk.Tx) ([]byte, error) {
+	bz, err := app.txEncoder(tx)
+	if err != nil {
+		return nil, errors.WithSecondaryError(ErrTxEncode, err)
+	}
+
+	_, _, _, _, err = app.runTx(runTxPrepareProposal, bz)
+	return bz, err
+}
```

Comment:
Returning a non-nil error canonically means other return values are invalid and should be ignored by the caller. So if you want to distinguish tx encoding errors from tx validation errors, that should be done with distinguishable error values.

```suggestion
// PrepareProposalVerifyTx performs transaction verification when a proposer is
// creating a block proposal during PrepareProposal. Any state committed to the
// PrepareProposal state internally will be discarded.
func (app *BaseApp) PrepareProposalVerifyTx(tx sdk.Tx) ([]byte, error) {
	bz, err := app.txEncoder(tx)
	if err != nil {
		return nil, errors.WithSecondaryError(ErrTxEncode, err)
	}

	if _, _, _, _, err := app.runTx(runTxPrepareProposal, bz); err != nil {
		return nil, errors.WithSecondaryError(ErrTxExecution, err)
	}
	
	return bz, nil
}
```

Reply:
No, I think this approach is fine IMO. The caller really only cares if bz == nil or not or if err == nil or not. I added the typed error, but it isn't necessary.

Reply:
By convention, if err == nil, bz is valid and can be evaluated by callers; if err != nil, bz is not valid and should not be evaluated by callers. Obviously you can write your API however you like, but it's highly non-idiomatic to return a valid []byte with a non-nil error.

Reply:
Yes, that is idiomatic. I was trying to avoid a few lines of code.